src/main/java/com/pusher/client/channel/impl/ChannelManager.java

```diff
@@ -137,6 +152,12 @@ public void run() {
 
                 if (connection.getState() == ConnectionState.CONNECTED) {
                     try {
+
+                        // only private encrypted channels can prepare themselves atm
+                        if (channel instanceof PrivateEncryptedChannelImpl) {
+                            ((PrivateEncryptedChannelImpl)channel).prepareChannel();
```

Comment:
No potential race-condition but other points from [this bit of feedback](https://github.com/pusher/pusher-websocket-java/pull/234/files#r399289302) apply here too. I don't think there's a need to require the client code to call a separate method here, all that logic could be trigged form the existing `toSubscribeMessage`.

Reply:
Danielle and I discussed this and decided to separate them because it was pretty surprising `toSubscribeMessage` having side effects, but as stated here: https://github.com/pusher/pusher-websocket-java/pull/234#discussion_r399116006, I believe it is worth promoting `prepare` and `teardown` as "hooks" in the base class and implementing for the other three so that there is no dependency on extra implementation details here.

Reply:
Separating those responsibilities internally here within the added implementation is fine. Changing the interface affecting 4 classes within an already complex interface/implementation hierarchy is a broader refactoring which should be given proper design thought (and feels outside of the very narrow timeframe we have now). For example `tearDown` is only needed in 1 out of 4 concrete implementations. Likewise `setUp` is not needed in every implementation.

The name `channel.toSubscribeMesssage` is strange no matter if and what side-effects it has, maybe a simple improvement for now could be a reasonable way forward at this stage: `prepareSubscribeMessage`. That would convey that more work is needed as opposed to just simple local in-mem message composition (mapping).

Reply:
After catching up on a call with Marek on Friday, I refactored out all the checks in the ChannelManager for `instanceof PrivateEncryptedChannelImpl` - the `prepare` happens at the start of `toSubscribeMessage` and the `tearDown` happens when the `updateState` is `UNSUBSCRIBED`.